### ISSUE TITLE
2023年4月13日

### ISSUE THREAD

**Issue by ssfc** [ssfc/wangdao#1151]
81	什么是元编程？chatGPT: 元编程是指在程序运行时，动态地创建、修改、或者操作程序的方式。这种方式可以使得程序更加灵活，更容易适应不同的需求。元编程可以通过反射、装饰器、元类等机制实现。在 Python 中，元编程的应用非常广泛，比如可以使用装饰器来对函数或类进行修饰，使用元类来动态创建类对象，使用反射来查看和修改对象的属性和方法等等。元编程可以帮助程序员在保持代码简洁和可维护性的前提下，实现更高级别的抽象和功能。(2023年3月28日)
82	是否等价？chatGPT: 是的，这两段代码是等价的。在 PyTorch 中，可以使用 to() 方法将张量从一个设备移动到另一个设备。也可以使用 device 参数在创建张量时指定设备。两者的效果是相同的。(2023年3月29日)
83	python中怎样判断哪些变量是被赋值成函数的变量？chatGPT: 在 Python 中，可以通过 callable() 函数来判断一个变量是否可调用，即是否可以作为函数进行调用。(2023年3月29日)
84	optimizer.zero_grad()什么意思？chatGPT: 在使用梯度下降算法进行优化时，我们需要在每次迭代时更新模型参数的梯度，以便最小化损失函数。在 PyTorch 中，这个过程包括三个步骤：计算损失函数的梯度，使用这些梯度来更新模型参数，然后将梯度清零，以便下一个迭代中使用。// optimizer.zero_grad() 这一行代码的作用就是将模型参数的梯度清零，以便在下一个迭代中使用新的梯度。这是因为 PyTorch 默认会在 backward() 函数中累积梯度，而不是覆盖它们。所以在每次迭代前需要显式地将梯度清零。(2023年3月29日)
85	chatGPT: 在 PyTorch 中，梯度下降包括三个步骤：计算损失函数的梯度，使用这些梯度来更新模型参数，然后将梯度清零，以便下一个迭代中使用。(2023年3月29日)
86	loss.backward()  什么意思？chatGPT: loss.backward()是PyTorch中的一个函数，用于计算反向传播。在训练神经网络时，我们需要根据当前的模型输出、真实标签以及损失函数来计算损失值，并通过反向传播算法计算模型参数的梯度。loss.backward()函数就是用来计算梯度的，它会自动地沿着计算图执行反向传播算法，将梯度值计算出来并保存在每个参数的.grad属性中，以便后续的优化步骤使用。(2023年3月29日)
87	optimizer.step()什么意思？chatGPT: optimizer.step()是优化器的一个方法，用于更新模型参数，使其朝着梯度下降方向移动。一般来说，我们需要在使用backward()方法计算完梯度之后，再调用optimizer.step()方法来更新参数。(2023年3月29日)
88	python list, numpy array和tensor有什么异同？chatGPT: Python中的列表（list）、NumPy数组（numpy array）和PyTorch张量（tensor）都是存储多个值的容器类型，它们的异同点如下：(1) list是Python的内置数据类型，可以存储不同类型的数据，而numpy array和tensor只能存储同一种数据类型的数据（通常为数值类型, int、float、bool）。(2) numpy array和tensor都可以进行高效的数值运算，因为它们底层使用的是连续的内存块，支持向量化运算，比list的循环计算要快得多。(3)  tensor是PyTorch的核心数据类型，是进行深度学习模型计算的基础，支持自动求导和GPU加速等特性，而list和numpy array则没有这些特性。(3) numpy array和tensor都提供了丰富的数学和统计函数，比如矩阵乘法、逆、特征值分解等，而list则没有这些函数。(4) list可以使用Python内置的一些函数和方法进行操作，如append、extend、sort等，而数组和张量则需要使用NumPy或PyTorch提供的函数和方法进行操作。(5) np array和tensor支持广播（broadcasting）机制，可以进行形状不同的数组和张量之间的计算，而列表不支持广播机制。// In conclusion，list适用于存储不同类型的数据和基本的数据操作，而np.array和tensor适用于数值计算和深度学习模型的计算。在深度学习中，tensor是最常用的容器类型。(2023年3月29日)
89	Comment: 结合以上观点, 吾人在计算两个word的距离的时候, 或者聚类的时候, 应该使用tensor, 而不是list或者np.array; (2023年3月29日)
90	pytorch怎样计算两个tensor之间的距离？chatGPT: (1) dist = torch.sum(torch.abs(a-b))  (2) dist = torch.sqrt(torch.sum(torch.pow(a-b, 2)))  (3) cos_sim = torch.sum(a*b) / (torch.sqrt(torch.sum(torch.pow(a, 2))) * torch.sqrt(torch.sum(torch.pow(b, 2))))  (2023年3月29日)
91	python能不能直接读取sql文件？chatGPT: Python本身不能直接读取SQL文件。通常需要使用Python中的库来连接到数据库，然后从数据库中读取数据。一些Python的库用于连接到数据库，例如sqlite3，psycopg2，pymysql等。你可以使用这些库中的函数来执行SQL查询并将结果读入Python。(2023年3月29日)
92	什么是phpmyadmin? chatGPT: phpMyAdmin是一个开源的用于管理MySQL数据库的免费软件，可以通过基于Web的用户界面执行各种数据库管理任务，例如创建、删除和修改数据库、表、列和索引等。它提供了一个易于使用的图形用户界面，使用户可以通过网页浏览器轻松管理MySQL数据库。phpMyAdmin支持多种操作系统和语言，是一个广泛使用的MySQL数据库管理工具。(2023年3月29日)
93	Comment: 过去吾人把sentence按类别分为method, relation, definition, others; 但是结合GPT3.5等大模型的成功来看, 没有必要做细分, 全放在一起一锅乱炖才是好的; (2023年3月29日)
94	Comment: 我认为, 训练所需的corpus.txt和word_id_number_symbol对应这种事应该有一个共同的起点, 那就是某个sql文件; 这就要求, 在执行export_sentence操作时, 也必须执行export database操作, 让exported corpus和exported sql file能够一一对应; 但是反过来却不一定需要, 因为export database的时候可能只是需要备份而已; DONE; (2023年3月29日)
95	mysql怎样显示一个table的结构？chatGPT: describle table_name; (2023年3月29日)
96	Output a txt file, containing word_name and word_id; 根据上一个Comment共同起点原则, 这个功能应该放在export sentence中一起输出; done; (2023年3月29日)
97	AI中有哪些常用的概率统计概念？chatGPT: 在人工智能中，常用的概率统计概念包括：   概率分布：描述一个随机变量在各个取值下出现的概率。常见的概率分布包括正态分布、均匀分布、泊松分布等。// 期望：是随机变量取值的加权平均值，反映了随机变量的平均水平。对于离散型随机变量，期望可以表示为每个取值的概率乘以对应取值的权值之和；对于连续型随机变量，期望可以表示为概率密度函数和对应的权值的积分。// 方差：是随机变量离其期望的平均距离的平方，反映了随机变量的离散程度。// 协方差：衡量两个随机变量之间的线性相关性。// 相关系数：是协方差除以两个随机变量各自的标准差的乘积，用于衡量两个随机变量之间的相关程度。// 最大似然估计：是一种常用的参数估计方法，用于在给定观测数据的情况下估计参数的值。// 贝叶斯估计：是一种基于贝叶斯定理的参数估计方法，将参数视为随机变量，通过先验分布和观测数据得到后验分布来估计参数。// 信息熵：是衡量信息不确定性的度量，用于衡量信息的平均信息量。// 互信息：衡量两个随机变量之间的相互依赖程度，是信息论中的一个重要概念。// 条件概率：表示在给定某个条件下另一个事件发生的概率，是贝叶斯定理中的重要组成部分。(2023年3月29日)
98	Rename parameter_save_root result_save_root; done; (2023年3月29日)
99	line = line.strip()是什么意思？chatGPT: 在Python中，字符串的strip()方法可以去除字符串开头和结尾处的空白符（包括空格、制表符、换行符等），返回一个新的字符串。在这个代码中，line代表从文件中读取的一行字符串，使用strip()方法可以去除这一行开头和结尾处的空白符，得到一个去除空白符后的新字符串。这是因为在读取文件时，有时会出现每一行的结尾都带有换行符\n或者其他空白符，如果不去除这些空白符，可能会导致后续处理出现错误。(2023年3月29日)
100	原因找到了, 是因为第二列数据有些也使用space作为分隔符; (2023年3月29日)
101	已知txt文件中第一列为word_id, 第二列为word_name, 如果第二列word_name中包含空格，两列数据用什么做分隔符比较好？chatGPT: 如果第二列word_name中可能包含空格，建议使用制表符\t作为分隔符，因为制表符在文本中不常见，而且可以在多数编辑器和解析器中被轻松地识别和处理。(2023年3月29日)
102	Comment: 用jetbrain家的phpstorm编辑PCF文件比vscode更好, 因为规范更多而且可以调试php; (2023年3月29日)
103	Use \t as separator in php; done; read it; done; (2023年3月29日)
104	Change turple to dict; done; evaluate it; done; (2023年3月29日)
105	组装椅子; 娘希匹太累了; (2023年3月30日)
106	\t \n等符号在c++, java, python等语言是通用的吗？为什么？chatGPT: \t, \n, \r, \b, \f, 等符号在C++, Java, Python等语言中都是通用的，因为它们都是被视为转义字符，用于表示一些不可见或者难以输入的字符。这些符号代表的含义在这些语言中也是相同的，如\t表示一个制表符，\n表示一个换行符，\r表示一个回车符，\b表示一个退格符，\f表示一个换页符等。不同语言之间可能存在一些细微的差别，比如在Python中，\r\n表示换行符，而在C++和Java中，\n表示换行符。但是，这些符号在不同语言中的基本含义是相同的。(2023年3月30日)
107	Create a mapping from vocab index to word id; done; (2023年3月30日)
108	Evaluate mapping; done; 确实一一对应; (2023年3月30日)
109	word_to_ix和word2ix哪种命名方式更好？或者你是否有更好的命名建议？chatGPT: 两种命名方式都可以，都符合Python的命名规范。一般来说，建议使用下划线命名法，即word_to_ix这种方式，因为这种方式更易读，而且在Python社区中也更为常见。不过命名也有一定的个人喜好和实际场景的考虑，可以根据具体情况灵活选择。(2023年3月30日)
110	这是一个很诡异的事情, 为什么word_to_ix和ix_to_word每次运行的结果都不一样？如果每次不一样, 这意味着储存的ix_to_word在test的时候不能通用; 查看vocab每次运行的结果是否相同; 卧槽, vocab每次运行的结果也不相同; (2023年3月30日)
111	vocab = set(all_tokens)每次执行时，为什么vocab的顺序每次都不一样？怎么才能让这个结果能否重复？chatGPT: 这是由于Python中的集合(set)是无序的，每次执行时元素的顺序可能不同，因此构建基于集合的词汇表(vocab)时，其顺序也可能不同。如果需要保持固定的顺序，可以使用Python中的有序集合，例如OrderedSet。(2023年3月30日)
112	vocab = collections.OrderedDict.fromkeys(all_tokens)是什么意思？chatGPT: collections.OrderedDict.fromkeys()是一个类方法，用于创建一个新的有序字典，其键来自于指定的可迭代对象，而值全部被设置为None。在这里，all_tokens是一个可迭代对象，所以我们可以使用这个方法创建一个有序字典vocab，其中all_tokens中的每个元素都被用作字典的键，而相应的值则被设置为None。// 例如，如果all_tokens是['apple', 'banana', 'apple']，那么执行vocab = collections.OrderedDict.fromkeys(all_tokens)将会创建一个有序字典vocab，其中键'apple'和'banana'出现在与它们在all_tokens中出现的相同的顺序，而值全部为None。// 需要注意的是，collections.OrderedDict相比于普通的字典，会在其内部维护元素的顺序，因此对于有序数据的处理具有一定的优势。(2023年3月30日)
113	怎样把vocab = collections.OrderedDict.fromkeys(all_tokens)按顺序建立word_to_ix和ix_to_word？chatGPT: word_to_ix = {word: i for i, (word, _) in enumerate(vocab.items())} (2023年3月30日)
114	Comment: 所有在test里面测试的项目, 在train结束的时候也应该做一遍, 以保证两个一致; (2023年3月30日)
115	Add word_name dict after train; done; (2023年3月30日)
116	并列if中重复的代码应该尽量抽取出去吗？chatGPT: 是的，如果有多个 if 语句中出现了相同的代码，应该尽量将它们抽取出来避免代码重复。这不仅可以使代码更加简洁，还可以减少代码中的错误和维护成本。可以将这些重复代码抽象成一个函数或者一个独立的代码块，在需要使用的地方进行调用即可。(2023年3月30日)
117	Extract word_id_name; done; (2023年3月30日)
118	Rename model in train_if as train_model; done; rename model in test if as test_model; done; (2023年3月30日)
119	Write func cluster; done; apply it in train if; done; apply it in test if; done; (2023年3月30日)
